release/release-offline.ts

```diff
@@ -0,0 +1,410 @@
+  );
+}
+
+if (!isValidCommitHash(commit)) {
+  error("You must provide a valid commit hash as the second argument");
+}
+
+if (!step) {
+  error("You must provide a step argument like --build or --publish");
+}
+
+/**************************************************
+          HELPERS
+ **************************************************/
+
+// mostly for type checking
+function getGithubCredentials() {
+  if (!GITHUB_TOKEN || !GITHUB_OWNER || !GITHUB_REPO) {
+    error("You must provide all github environment variables in .env-template");
+    process.exit(1);
+  }
+
+  return { GITHUB_TOKEN, GITHUB_OWNER, GITHUB_REPO };
+}
+
+async function checkReleased() {
+  const { GITHUB_OWNER, GITHUB_REPO } = getGithubCredentials();
+
+  const hasThisVersionBeenReleased = await hasBeenReleased({
+    github,
+    owner: GITHUB_OWNER,
+    repo: GITHUB_REPO,
+    version,
+  });
+
+  if (hasThisVersionBeenReleased) {
+    error(`Version ${version} has already been released`);
+  }
+}
+
+async function checkJar() {
+  // check if built jar exists
+  const requiredFiles = [
+    `${JAR_PATH}/metabase.jar`,
+    `${JAR_PATH}/COMMIT-ID`,
+    `${JAR_PATH}/SHA256.sum`,
+  ];
+
+  requiredFiles.forEach(file => {
+    if (!fs.existsSync(file)) {
+      error(`You must build the jar first. ${file} does not exist`);
+    }
+  });
+
+  const versionProperties = (
+    await $`jar xf ${JAR_PATH}/metabase.jar version.properties && cat version.properties`
+  ).toString();
+
+  log(versionProperties, "green");
+
+  if (!versionProperties.includes(`tag=${version}`)) {
+    error(`This jar does not match the input version: ${version}`);
+  }
+}
+
+// eslint-disable-next-line no-console
+const log = (message, color = "blue") =>
+  console.log(chalk[color](`\n${message}\n`));
+
+function error(message) {
+  log(`⚠️   ${message}`, "red");
+  process.exit(1);
+}
+
+/**************************************************
+          BUILD STEP
+ **************************************************/
+async function build() {
+  log(`🚀 Building ${edition} jar for ${version} from commit ${commit}`);
+
+  // check build environment
+  const majorVersion = Number(getMajorVersion(version));
+  const nodeVersion = (await $`node --version`).toString();
+  const javaVersion = (await $`java -version`).toString();
+
+  if (!nodeVersion.includes(`v${versionRequirements[majorVersion].node}`)) {
+    error(
+      `Node version must be v${versionRequirements[majorVersion].node}, you are running ${nodeVersion}`,
+    );
+  }
+
+  if (
+    !javaVersion.includes(
+      `openjdk version "${versionRequirements[majorVersion].java}`,
+    )
+  ) {
+    error(
+      `Java version must be ${versionRequirements[majorVersion].java}, you are running ${javaVersion}`,
+    );
+  }
+
+  const currentBranch = (await $`git branch --show-current`).toString().trim();
+
+  try {
+    await $`git fetch --all`;
+    await $`git stash && git checkout ${commit}`;
```

Comment:
Instead of stashing, can we throw an error early on in the script if `git status` tells us there are any changes?

Reply:
👍

![Screen Shot 2023-11-09 at 4 23 18 PM](https://github.com/metabase/metabase/assets/30528226/87cada41-5925-4aad-a9be-43136945a091)